### DISCUSSION TITLE
How to retry a failed step in github actions workflow?

### DISCUSSION THREAD

**Discussion by riprasad** [community/community#27121] [Actions]
<p>Hi Team,</p>
<p>I am aware that we have an option to re-run the failed workflow by clicking the button <strong>Re-run checks</strong> .  But that would re-run the entire workflow again. I am also aware that currently we cannot resume the workflow directly from the failed step. But, does GitHub Actions provides a way to automatically retry the failed steps?</p>
<p>say for example, if my <strong>maven deploy</strong> step fails due to network issues for the first time, it should retry, let’s say 3 times, before it actually fails. Do we have a retry mechanism available ?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/riprasad">@riprasad</a>,</p>
<p>Thank you for reaching this out!</p>
<p>Currently it’s not supported to retry the failed step but only the workflow, you can raise a feature_request <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a> where github product manager will take a review.</p>
<p>As a workaround, you can use below sample code for the retry, please refer to my comment:</p>
<pre><code class="lang-auto">    steps:
      - name: test
        continue-on-error: true          # make the step always success
        id: test
        run: |
          exit 1                         # let's say 1st failed. for you, here could be mvn deploy command.
      - name: retry
        continue-on-error: true
        id: retry1
        if: steps.test.outcome=='failure'         # check the step outcome, retry 1st time
        run: |
          exit 1                                # if it fails again. if it's succeeds, the latter steps will be skipped, and status is successful.
      - name: retry 2
        continue-on-error: true
        id: retry2
        if: steps.retry1.outcome=='failure'   # check the step outcome, retry 2nd time
        run: |
          echo retry 2nd time
          exit 1                                            # fail again
      - name: retry 3
        continue-on-error: true
        id: retry3
        if: steps.retry2.outcome=='failure'      # check the step outcome, retry 3rd time
        run: |
          echo retry 3rd time                 # it passed in retry 3
      - name: set the status              # set the workflow status if command failed
        if: always()
        run: |
          if ${{ steps.test.outcome=='success' || steps.retry1.outcome=='success' || steps.retry2.outcome=='success' || steps.retry3.outcome=='success' }}; then
             echo fine
          else
             exit 1
          fi
</code></pre>
<p>set <code>continue-on-error: true</code> will always make the step success, however <code>steps.{stepid}.outcome</code> can get the command result, hence at last, add a step to show the real execution status, it will fail if all(test, retry1, retry2, retry3) steps fails, but show passed if there’s a successful step.</p>
<p>Please refer to my workflow for more details.<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/weide-zhou/ticket13/runs/944172248?check_suite_focus=true" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/52772761/181101785-32c5bf49-f240-470a-bef7-ff905bb8e151.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/weide-zhou/ticket13/runs/944172248?check_suite_focus=true" target="_blank" rel="nofollow noopener">weide-zhou/ticket13</a></h3>

<p>Contribute to weide-zhou/ticket13 development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Thanks.</p>

**Comment by webbertakken** [comment]
<p>Related: <a href="https://github.community/t/retry-for-failed-steps/17136" class="inline-onebox">Retry for failed steps</a></p>

**Reply by knocte** [reply]
@webbertakken broken link

**Reply by webbertakken** [reply]
Thanks for reporting the broken link. Unfortunately I'm unable to search for the ID of that thread it seems. I can also not find it by the name of the thread.

**Comment by riprasad** [comment, marked as answer]
<p>I was able to achieve retries by using this simple workaround</p>
<pre><code class="lang-auto">- name: Maven Deploy
  run: |
    # Retry 3 times before the steps actually fails
    (echo "===== Maven Deploy Attempt: 1 ====" &amp;&amp; mvn deploy) || \
    (echo "===== Maven Deploy Attempt: 2 ====" &amp;&amp; mvn deploy) || \
    (echo "===== Maven Deploy Attempt: 3 ====" &amp;&amp; mvn deploy) || \
    (echo "==== Maven Deploy Step Failed ====" &amp;&amp; exit 1)
</code></pre>
